server/src/main/java/com/saralein/server/request/RequestParser.java

```diff
@@ -1,69 +1,61 @@
 package com.saralein.server.request;
 
-import static com.saralein.server.Constants.CRLF;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
+import static com.saralein.server.Constants.CRLF;
 
 public class RequestParser {
-    public Request parse(String request) throws Exception {
+    public Request parse(String rawRequest) throws Exception {
         try {
-            List<String> fullRequest = Arrays.asList(split(request, CRLF));
-            HashMap<String, String> parsedRequest = createRequestMap(fullRequest);
-            return new Request(parsedRequest);
+            List<String> request = Arrays.asList(split(rawRequest, CRLF));
+            String[] requestLine = parseRequestLine(request);
+
+            return new Request.Builder()
+                    .addMethod(parseMethod(requestLine))
+                    .addUri(parseUri(requestLine))
+                    .addBody(parseBody(request))
+                    .addHeaders(parseHeaders(request))
+                    .build();
         } catch(ArrayIndexOutOfBoundsException e) {
             throw new Exception("Bad request. Connection closed.");
         }
     }
 
-    private HashMap<String, String> createRequestMap(List<String> fullRequest) {
-        HashMap<String, String> parsedRequest = addRequestLine(fullRequest);
-
-        if (fullRequest.size() > 0) {
-            parsedRequest.putAll(addHeadersAndBody(fullRequest));
-        }
-
-        return parsedRequest;
+    private String[] parseRequestLine(List<String> request) {
+        return split(request.get(0), " ");
     }
 
-    private HashMap<String, String> addRequestLine(List<String> fullRequest) {
-        String[] requestLine = splitRequestLine(fullRequest);
+    private String parseBody(List<String> request) {
+        for (String line: request) {
+            if (line.startsWith("body")) {
+                return split(line, ":")[1].trim();
+            }
+        }
```

Comment:
I've reworked this a bit (and other pieces of the parser as well).  I think there is always room for improvement in that parser.  One thing I'm not so sure about though: I cast a map to a hashmap in `parseHeaders` now.  I did this because I didn't want to make request `addHeaders` more generic.

Reply:
Why don't want to make it more generic?